Use pkg-config to get x11 and xt flags

zmughal opened the pull request:
This PR is on top of the commits of <https://github.com/djerius/pgplot-autotool/pull/2>.

As discussed in <https://github.com/djerius/pgplot-autotool/pull/2>, some
package managers (e.g., Homebrew) do not place x11 and xt files in the standard
locations. This uses `pkg-config` to get that information first.

zmughal committed:
Use pkg-config to find x11 and xt

zmughal committed:
Add *.o to .gitignore

zmughal committed:
GHA: Use same make step for all platforms

zmughal committed:
Look for modules to support -lSM -lICE

Because these are listed in `$X_PRE_LIBS`.

zmughal committed:
pkg-config: xt already depends on x11, so add once

zmughal committed:
GHA: Add debugging of pkg-config files

djerius commented on configure.ac:
```diff
@@ -73,6 +73,41 @@ dnl   automatically (see the automake manual's FAQ for justification).
+have_x11=no
+have_xt=no
+AS_IF([ test "x$no_x" = "x"],
+      [
```
I think there's something awry with the logic here.  `AC_PATH_XTRA` is invoked above, and, from the fine `autoconf` manual, section 5.11:

> If [finding X fails], or the user gave the command line option '--without-x', set the shell variable 'no_x' to 'yes'; otherwise set it to the empty string.

The true state of this `AS_IF` conditional is when `test "x$no_x" = "x"` succeeds, which requires that `$no_x` is the empty string and, from above,  `X` has already been found, so evaluating the body of the conditional is extra work.

Conversely, if `X` has not been found, `$no_x` is set to `yes`, but unfortunately that's also the value when the user explicitly doesn't want `X`, signified by setting the `--without-x` command line parameter.

I think the sense of the conditional is backwards, but that it won't work anyway because of the degeneracy between `AC_PATH_EXTRA` failing and the user setting `--without-x`

zmughal replied:
How about checking `$with_x` directly and use `pkg-config` as an additional search method to `AC_PATH_XTRA`?

zmughal replied:
I have added a commit below that does what I suggested by using `$with_x`.

zmughal committed:
GHA: add builds to check disabling X11

zmughal commented:
I have rebased on top of the main branch.

I also added a commit to test using the `--without-x` option.

zmughal committed:
Use pkg-config for x11 in addition to AC_PATH_XTRA

oodler577 commented on .github/workflows/ci.yml:
```diff
@@ -30,13 +34,12 @@ jobs:
+      - name: Show pkg-config files
         run: |
-          make LDFLAGS="$(pkg-config --libs --static libpng xt)"
+          head -n100 *.pc
```
Is this `head` just for output? Assuming so since you're applying it to anything with a `.pc` extension.

zmughal replied:
Yes, that's just debugging output since I couldn't find a target to use for `make test`. Should also help make sure that disabling X11 means the `pkg-config` does not have a `Requires:` with `x11` in the list.

oodler577 reviewed (commented):
defering judgement; looked at this mainly for my own edification and if I could see something obvious :)

zmughal commented:
@djerius, have a moment to look over this?

djerius commented on configure.ac:
```diff
@@ -73,6 +73,42 @@ dnl   automatically (see the automake manual's FAQ for justification).
 dnl   Not all platforms have X
+have_x11=no
+have_xt=no
+AS_IF([ test "$with_x" != no ],
```
```suggestion
dnl	+--------+------------------------+----------+----------+--------+-----+
dnl	|Variable|      AC_PATH_XTRA      |-without-x|-with-x=no|-with-x | --x |
dnl	+--------+------------+-----------+----------+----------+--------+-----+
dnl	|        |  success   |  failure  |          |          |        |     |
dnl	+--------+------------+-----------+----------+----------+--------+-----+
dnl	|  no_x  |     ''     |   'yes'   |  'yes'   |          |        |     |
dnl	+--------+------------+-----------+----------+----------+--------+-----+
dnl	| with_x |            |           |   'no'   |   'no'   | 'yes'  |'yes'|
dnl	+--------+------------+-----------+----------+----------+--------+-----+
dnl	Notes:
dnl     1) AC_PATH_EXTRA runs first and modifies no_x in the presence of -without-x
dnl     2) If --with-x is not specified, AC_PATH_EXTRA assumes with_x='yes'
dnl
dnl     Check pkg-config if the user did not forbid X and AC_PATH_EXTRA failed to find it
AS_IF([ test "x$with_x" != xno -a "x$no_x" != x ],

```

zmughal replied:
I'll add this as a commit and see if it works with Homebrew's configuration.

zmughal committed:
Test $no_x as well to detect AC_PATH_XTRA failure

Co-authored-by: djerius <djerius@cfa.harvard.edu>

zmughal committed:
Set macOS LIBRARY_PATH for Homebrew

Per the macOS ld(1) man page <https://www.unix.com/man-page/OSX/1/ld/>,

> The default library search path is /usr/lib then /usr/local/lib.

When `HOMEBREW_PREFIX=/usr/local`, this is where the `.dylib` files for
X11 are symlinked by Homebrew.

When `AC_PATH_XTRA` is called, it uses `/usr/bin/gcc` (Apple clang) and
since the libraries are found in `/usr/local/lib`, it does not need to
specify any `-L` paths.

However when `libtool` is used to call Homebrew `gfortran`, it sets
the flag `-syslibroot` which per the man page for ld(1):

> The -syslibroot option will prepend a prefix to all search paths.

and this means that it no longer looks under `/usr/local/lib` by
default. To fix this, add the default path back by setting the
`LIBRARY_PATH` environment variable.

zmughal commented:
@djerius, now that change should work. More details in the commit:

> Set macOS LIBRARY_PATH for Homebrew

zmughal committed:
GHA: Test Homebrew linking via pkg-config only

This is to test that if libraries are not symlinked into the Homebrew
prefix (here by using `brew unlink`), then `autotool` will use
`pkg-config` to find the libraries.

zmughal committed:
Mark when X11 is not found via pkg-config

zmughal committed:
Add Homebrew formulae to PKG_CONFIG_PATH

zmughal commented:
@djerius, the latest commits should now test the case where `AC_PATH_XTRA` fails and must fallback on `pkg-config`. I also added the extra parameter to `PKG_CHECK_MODULES` so that it does not error out of the script if `pkg-config` can not find `x11`.

djerius commented:
Sorry about the delay!

zmughal commented:
Not a problem!

I was wondering if the next thing to tackle is getting #1 incorporated (hopefully it can be done and the license can be worked out).
